resources/regexes.ts

```diff
@@ -131,26 +203,88 @@ const parseHelper = <T extends LogDefinitionTypes>(
 
     str += separator;
 
-    const value = fields[keyStr];
-    if (typeof value !== 'object')
-      throw new Error(`${defKey}: invalid value: ${JSON.stringify(value)}`);
+    if (typeof parseField !== 'object')
+      throw new Error(`${defKey}: invalid value: ${JSON.stringify(parseField)}`);
 
     const fieldDefault = fieldName !== undefined && fieldsWithPotentialColons.includes(fieldName)
       ? matchWithColonsDefault
       : matchDefault;
-    const fieldValue = fields[keyStr]?.value?.toString() ?? fieldDefault;
-
-    if (fieldName !== undefined) {
-      str += Regexes.maybeCapture(
-        // more accurate type instead of `as` cast
-        // maybe this function needs a refactoring
-        capture,
-        fieldName,
-        params[fieldName],
-        fieldValue,
-      );
+    const defaultFieldValue = parseField.value?.toString() ?? fieldDefault;
+    const fieldValue = params[fieldName];
+
+    if (isRepeatingField(fields[keyStr]?.repeating, fieldValue)) {
+      let repeatingArray: RepeatingFieldsMap<T> = fieldValue;
+      const origRepeatingArray = repeatingArray;
+
+      const sortFn = fields[keyStr]?.sortFn;
+      // Allow sorting if needed
+      if (sortFn)
+        repeatingArray = [...repeatingArray].sort(sortFn);
+
+      repeatingArray.forEach((rep) => {
+        if (typeof rep === 'object' && rep !== null && Object.keys(rep).length > 0) {
+          let fieldRegex = '';
+          // Rather than looping over the keys defined on the object,
+          // loop over the base type def's keys. This enforces the correct order.
+          fields[keyStr]?.repeatingKeys?.forEach((key) => {
+            fieldRegex += separator;
+            if (!(key in rep)) {
+              // If we don't have a value for this key, insert a placeholder
```

Comment:
When would this happen? Sorry that I'm just not understanding this case.

Reply:
Hmm, I see.  This was unclear to me.  I think `.*` would be incorrect here, as it would match across multiple fields.  (I mean, arguably we could detect/fix this and it's already an issue for other regexes potentially.)  Can you leave a comment that this is why this case exists, and maybe also put a test case for this?  This was non-obvious to me.

> (Though this raises a question of "maybe we should just optionally match against all possible fields if it's a capture: true regex, even if that results in a really big and potentially slow regex?)

For me the answer is no.  I think if you want all of them, you just pull the line apart yourself.  In terms of match group names, I think it'd be hard to number them in a consistent way.

Reply:
I've got a local version right now that given this NetRegexes:
```ts
      NetRegexes.combatantMemory({
        pair: [{ key: 'IsCasting1', value: '1' }, { key: 'CastBuffID', value: '.*?' }],
      })
```
Produces this regex:
```ts
/^(?<type>261)\|(?<timestamp>[^|]*)\|(?<change>[^|]*)\|(?<id>[^|]*)\|(?:(?<AggressionStatuskey>AggressionStatus)\|(?<AggressionStatusvalue>[^|]*)\|)?(?:(?<BNpcIDkey>BNpcID)\|(?<BNpcIDvalue>[^|]*)\|)?(?:(?<BNpcNameIDkey>BNpcNameID)\|(?<BNpcNameIDvalue>[^|]*)\|)?(?:(?<CastBuffIDkey>CastBuffID)\|(?<CastBuffIDvalue>.*?)\|)(?:(?<CastDurationCurrentkey>CastDurationCurrent)\|(?<CastDurationCurrentvalue>[^|]*)\|)?(?:(?<CastDurationMaxkey>CastDurationMax)\|(?<CastDurationMaxvalue>[^|]*)\|)?(?:(?<CastTargetIDkey>CastTargetID)\|(?<CastTargetIDvalue>[^|]*)\|)?(?:(?<CurrentCPkey>CurrentCP)\|(?<CurrentCPvalue>[^|]*)\|)?(?:(?<CurrentGPkey>CurrentGP)\|(?<CurrentGPvalue>[^|]*)\|)?(?:(?<CurrentHPkey>CurrentHP)\|(?<CurrentHPvalue>[^|]*)\|)?(?:(?<CurrentMPkey>CurrentMP)\|(?<CurrentMPvalue>[^|]*)\|)?(?:(?<CurrentWorldIDkey>CurrentWorldID)\|(?<CurrentWorldIDvalue>[^|]*)\|)?(?:(?<Distancekey>Distance)\|(?<Distancevalue>[^|]*)\|)?(?:(?<EffectiveDistancekey>EffectiveDistance)\|(?<EffectiveDistancevalue>[^|]*)\|)?(?:(?<Headingkey>Heading)\|(?<Headingvalue>[^|]*)\|)?(?:(?<IDkey>ID)\|(?<IDvalue>[^|]*)\|)?(?:(?<IsCasting1key>IsCasting1)\|(?<IsCasting1value>1)\|)(?:(?<IsCasting2key>IsCasting2)\|(?<IsCasting2value>[^|]*)\|)?(?:(?<IsTargetablekey>IsTargetable)\|(?<IsTargetablevalue>[^|]*)\|)?(?:(?<Jobkey>Job)\|(?<Jobvalue>[^|]*)\|)?(?:(?<Levelkey>Level)\|(?<Levelvalue>[^|]*)\|)?(?:(?<MaxCPkey>MaxCP)\|(?<MaxCPvalue>[^|]*)\|)?(?:(?<MaxGPkey>MaxGP)\|(?<MaxGPvalue>[^|]*)\|)?(?:(?<MaxHPkey>MaxHP)\|(?<MaxHPvalue>[^|]*)\|)?(?:(?<MaxMPkey>MaxMP)\|(?<MaxMPvalue>[^|]*)\|)?(?:(?<ModelStatuskey>ModelStatus)\|(?<ModelStatusvalue>[^|]*)\|)?(?:(?<MonsterTypekey>MonsterType)\|(?<MonsterTypevalue>[^|]*)\|)?(?:(?<NPCTargetIDkey>NPCTargetID)\|(?<NPCTargetIDvalue>[^|]*)\|)?(?:(?<Namekey>Name)\|(?<Namevalue>[^|]*)\|)?(?:(?<OwnerIDkey>OwnerID)\|(?<OwnerIDvalue>[^|]*)\|)?(?:(?<PCTargetIDkey>PCTargetID)\|(?<PCTargetIDvalue>[^|]*)\|)?(?:(?<PartyTypekey>PartyType)\|(?<PartyTypevalue>[^|]*)\|)?(?:(?<PosXkey>PosX)\|(?<PosXvalue>[^|]*)\|)?(?:(?<PosYkey>PosY)\|(?<PosYvalue>[^|]*)\|)?(?:(?<PosZkey>PosZ)\|(?<PosZvalue>[^|]*)\|)?(?:(?<Radiuskey>Radius)\|(?<Radiusvalue>[^|]*)\|)?(?:(?<Statuskey>Status)\|(?<Statusvalue>[^|]*)\|)?(?:(?<TargetIDkey>TargetID)\|(?<TargetIDvalue>[^|]*)\|)?(?:(?<TransformationIdkey>TransformationId)\|(?<TransformationIdvalue>[^|]*)\|)?(?:(?<Typekey>Type)\|(?<Typevalue>[^|]*)\|)?(?:(?<WeaponIdkey>WeaponId)\|(?<WeaponIdvalue>[^|]*)\|)?(?:(?<WorldIDkey>WorldID)\|(?<WorldIDvalue>[^|]*)\|)?(?:(?<WorldNamekey>WorldName)\|(?<WorldNamevalue>[^|]*)\|)?/i
```

Example screenshot showing regex with highlighting and matching regex:
![image](https://github.com/quisquous/cactbot/assets/6119598/4cfbb6fb-e4f0-4bc4-b769-e7def6aa8eed)

Example screenshot showing what `matches` would be in the context of a cactbot trigger function:
![image](https://github.com/quisquous/cactbot/assets/6119598/0416b6a0-cb13-4125-bf13-c71539ad3c88)

I think this makes a lot more sense, and because there's no `.*?` this should also be significantly faster than the solution currently in this PR, since the regex engine should early out each optional capturing group when the `key` part of the line fails to match.

The only thing that doesn't sit well with me is the naming of the capturing groups, `IsCasting1key` and `IsCasting1value` might be a bit confusing and maybe they should be `IsCasting1__key` and `IsCasting1__value` instead?

(this also follows the `capture: true`/`capture: false` logic, so if `capture: false` it won't insert capturing groups)

Reply:
Ooooh, nice.  One observation from me is why do we need a key/value pair for the keys, when the key/value pair for the value has the key already in it.  In other words, `IsCasting: '1',` seems sufficient, and you don't need `IsCastingKey: 'IsCasting', IsCastingValue: '1',`.

Reply:
My concern is conflicting fields, e.g. there's a `type` by default in all lines (`261` in this case), and there's a `type` on `PluginCombatantState` (which this has renamed to `Type` but still).

So I'd be fine with e.g. `matches['pair.Type']` maybe? Or `matches.pair_Type`, or some other variation on that.

Reply:
(but yes, I agree that having the key as a property on `matches` is redundant)

Reply:
`pairType`  is a bit more camelcase-y on the eyes and wouldn't need the square brackets to access.

Reply:
Alright, this look good to you? I'll see if I can get some actual type hinting on matches as well but that's a bit more iffy.

```ts
/^(?<type>261)\|(?<timestamp>[^|]*)\|(?<change>[^|]*)\|(?<id>[^|]*)\|(?:AggressionStatus\|(?<pairAggressionStatus>[^|]*)\|)?(?:BNpcID\|(?<pairBNpcID>[^|]*)\|)?(?:BNpcNameID\|(?<pairBNpcNameID>[^|]*)\|)?(?:CastBuffID\|(?<pairCastBuffID>.*?)\|)(?:CastDurationCurrent\|(?<pairCastDurationCurrent>[^|]*)\|)?(?:CastDurationMax\|(?<pairCastDurationMax>[^|]*)\|)?(?:CastTargetID\|(?<pairCastTargetID>[^|]*)\|)?(?:CurrentCP\|(?<pairCurrentCP>[^|]*)\|)?(?:CurrentGP\|(?<pairCurrentGP>[^|]*)\|)?(?:CurrentHP\|(?<pairCurrentHP>[^|]*)\|)?(?:CurrentMP\|(?<pairCurrentMP>[^|]*)\|)?(?:CurrentWorldID\|(?<pairCurrentWorldID>[^|]*)\|)?(?:Distance\|(?<pairDistance>[^|]*)\|)?(?:EffectiveDistance\|(?<pairEffectiveDistance>[^|]*)\|)?(?:Heading\|(?<pairHeading>[^|]*)\|)?(?:ID\|(?<pairID>[^|]*)\|)?(?:IsCasting1\|(?<pairIsCasting1>1)\|)(?:IsCasting2\|(?<pairIsCasting2>[^|]*)\|)?(?:IsTargetable\|(?<pairIsTargetable>[^|]*)\|)?(?:Job\|(?<pairJob>[^|]*)\|)?(?:Level\|(?<pairLevel>[^|]*)\|)?(?:MaxCP\|(?<pairMaxCP>[^|]*)\|)?(?:MaxGP\|(?<pairMaxGP>[^|]*)\|)?(?:MaxHP\|(?<pairMaxHP>[^|]*)\|)?(?:MaxMP\|(?<pairMaxMP>[^|]*)\|)?(?:ModelStatus\|(?<pairModelStatus>[^|]*)\|)?(?:MonsterType\|(?<pairMonsterType>[^|]*)\|)?(?:NPCTargetID\|(?<pairNPCTargetID>[^|]*)\|)?(?:Name\|(?<pairName>[^|]*)\|)?(?:OwnerID\|(?<pairOwnerID>[^|]*)\|)?(?:PCTargetID\|(?<pairPCTargetID>[^|]*)\|)?(?:PartyType\|(?<pairPartyType>[^|]*)\|)?(?:PosX\|(?<pairPosX>[^|]*)\|)?(?:PosY\|(?<pairPosY>[^|]*)\|)?(?:PosZ\|(?<pairPosZ>[^|]*)\|)?(?:Radius\|(?<pairRadius>[^|]*)\|)?(?:Status\|(?<pairStatus>[^|]*)\|)?(?:TargetID\|(?<pairTargetID>[^|]*)\|)?(?:TransformationId\|(?<pairTransformationId>[^|]*)\|)?(?:Type\|(?<pairType>[^|]*)\|)?(?:WeaponId\|(?<pairWeaponId>[^|]*)\|)?(?:WorldID\|(?<pairWorldID>[^|]*)\|)?(?:WorldName\|(?<pairWorldName>[^|]*)\|)?/i
```

![image](https://github.com/quisquous/cactbot/assets/6119598/5f4123b6-2a91-4fda-ab22-4e99f8d7ebd8)

Reply:
Yeah, seems reasonable.  Seems wordy, but it's definitely better than trying to special-case `type` or having `type` vs `Type`, and I can't come up with anything better.

Reply:
Success! Just need to do a bit more cleanup and this'll be ready.

![image](https://github.com/quisquous/cactbot/assets/6119598/b785e939-ec37-401c-87bf-82554afa20b0)